sw/device/lib/dif/dif_alert_handler.c

```diff
@@ -4,9 +4,930 @@
+static bool classify_local_alerts(
+    const dif_alert_handler_t *handler,
+    const dif_alert_handler_class_config_t *class) {
+  if (class->local_alerts == NULL && class->local_alerts_len != 0) {
+    return false;
+  }
+
+  uint32_t enable_reg = mmio_region_read32(
+      handler->params.base_addr, ALERT_HANDLER_LOC_ALERT_EN_REG_OFFSET);
+  uint32_t alerts_reg = mmio_region_read32(
+      handler->params.base_addr, ALERT_HANDLER_LOC_ALERT_CLASS_REG_OFFSET);
+
+  for (int i = 0; i < class->alerts_len; ++i) {
+    uint32_t classification;
+    switch (class->alert_class) {
+      case kDifAlertHandlerClassA:
+        classification = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_CLASSA;
+        break;
+      case kDifAlertHandlerClassB:
+        classification = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_CLASSB;
+        break;
+      case kDifAlertHandlerClassC:
+        classification = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_CLASSC;
+        break;
+      case kDifAlertHandlerClassD:
+        classification = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_CLASSD;
+        break;
+      default:
+        return false;
+    }
+
+    bitfield_bit32_index_t enable_bit;
+    bitfield_field32_t field;
+    switch (class->local_alerts[i]) {
+      case kDifAlertHandlerLocalAlertAlertPingFail:
+        enable_bit = ALERT_HANDLER_LOC_ALERT_EN_EN_LA_0;
+        field.mask = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_MASK;
+        field.index = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_0_OFFSET;
+        break;
+      case kDifAlertHandlerLocalAlertEscalationPingFail:
+        enable_bit = ALERT_HANDLER_LOC_ALERT_EN_EN_LA_1;
+        field.mask = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_1_MASK;
+        field.index = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_1_OFFSET;
+        break;
+      case kDifAlertHandlerLocalAlertAlertIntegrityFail:
+        enable_bit = ALERT_HANDLER_LOC_ALERT_EN_EN_LA_2;
+        field.mask = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_2_MASK;
+        field.index = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_2_OFFSET;
+        break;
+      case kDifAlertHandlerLocalAlertEscalationIntegrityFail:
+        enable_bit = ALERT_HANDLER_LOC_ALERT_EN_EN_LA_3;
+        field.mask = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_3_MASK;
+        field.index = ALERT_HANDLER_LOC_ALERT_CLASS_CLASS_LA_3_OFFSET;
+        break;
+      default:
+        return false;
+    }
+
+    enable_reg = bitfield_bit32_write(enable_reg, enable_bit, true);
+    alerts_reg = bitfield_field32_write(alerts_reg, field, classification);
+  }
+
+  mmio_region_write32(handler->params.base_addr,
+                      ALERT_HANDLER_LOC_ALERT_EN_REG_OFFSET, enable_reg);
+  mmio_region_write32(handler->params.base_addr,
+                      ALERT_HANDLER_LOC_ALERT_CLASS_REG_OFFSET, alerts_reg);
+  return true;
+}
+
+/**
+ * Configures the control registers of a particular alert handler class.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool configure_class(const dif_alert_handler_t *handler,
```

Comment:
It almost feels like this function could be split in multiple static functions (matter of preference I suppose).

Reply:
Maybe, but I feel like that would break up the control flow even more. I'll add comments to deliniate the different parts of the function.

It just feels big because it has a lot of silly switching going on.